core/src/main/scala/kafka/server/FetcherEventManager.scala

```diff
@@ -0,0 +1,148 @@
+package kafka.server
+
+
+import java.util.concurrent.TimeUnit
+
+import com.yammer.metrics.core.Gauge
+import kafka.cluster.BrokerEndPoint
+import kafka.metrics.{KafkaMetricsGroup, KafkaTimer}
+import kafka.utils.ShutdownableThread
+import org.apache.kafka.common.internals.KafkaFutureImpl
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.common.{KafkaFuture, TopicPartition}
+
+import scala.collection.{Map, Set}
+
+trait FetcherEventProcessor {
+  def process(event: FetcherEvent)
+  def fetcherStats: AsyncFetcherStats
+  def fetcherLagStats : AsyncFetcherLagStats
+  def sourceBroker: BrokerEndPoint
+  def close(): Unit
+}
+
+object FetcherEventManager {
+  val EventQueueTimeMetricName = "EventQueueTimeMs"
+  val EventQueueSizeMetricName = "EventQueueSize"
+  val ScheduledEventQueueSizeMetricName = "ScheduledEventQueueSize"
+}
+
+/**
+ * The FetcherEventManager can spawn a FetcherEventThread, whose main job is to take events from a
+ * FetcherEventBus and executes them in a FetcherEventProcessor.
+ * @param name
+ * @param fetcherEventBus
+ * @param processor
+ * @param time
+ */
+class FetcherEventManager(name: String,
+                          fetcherEventBus: FetcherEventBus,
+                          processor: FetcherEventProcessor,
+                          time: Time) extends KafkaMetricsGroup {
+
+  import FetcherEventManager._
+
+  val rateAndTimeMetrics: Map[FetcherState, KafkaTimer] = FetcherState.values.flatMap { state =>
+    state.rateAndTimeMetricName.map { metricName =>
+      state -> new KafkaTimer(newTimer(metricName, TimeUnit.MILLISECONDS, TimeUnit.SECONDS))
+    }
+  }.toMap
+
+  @volatile private var _state: FetcherState = FetcherState.Idle
+  private[server] val thread = new FetcherEventThread(name)
+
+  def fetcherStats: AsyncFetcherStats = processor.fetcherStats
+  def fetcherLagStats : AsyncFetcherLagStats = processor.fetcherLagStats
+  def sourceBroker: BrokerEndPoint = processor.sourceBroker
+  def isThreadFailed: Boolean = thread.isThreadFailed
+
+  private val eventQueueTimeHist = newHistogram(EventQueueTimeMetricName)
+
+  newGauge(
+    EventQueueSizeMetricName,
+    new Gauge[Int] {
+      def value: Int = {
+        fetcherEventBus.eventQueueSize()
+      }
+    }
+  )
```

Comment:
Does it make more sense to trace the tail queue size?  Or is that done automatically with the kafka metrics?

Reply:
The metrics that support percentile statistics need to be a histogram (created via newHistogram), or a timer (created via newTimer). Internally, a yammer timer records durations using a histogram.
In essence, only the ones backed by a histogram support percentile stats, and we cannot support that when measuring the queue length using a gauge.

Reply:
Thanks for checking on this.  My bad I didn't understand our internal graphs are mislabeled.